drivers/timer/nrf_grtc_timer.c

```diff
@@ -555,18 +557,48 @@ void sys_clock_set_timeout(int32_t ticks, bool idle)
 		return;
 	}
 
-	ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : MIN(MAX_TICKS, MAX(ticks, 0));
+	uint32_t ch = system_clock_channel_data.channel;
+
+	if ((cc_value == expired_cc) && (ticks < MAX_REL_TICKS)) {
+		uint32_t cyc = ticks * CYC_PER_TICK;
+
+		if (cyc == 0) {
+			/* GRTC will expire anyway since HW ensures that past value triggers an
+			 * event but we need to ensure to always progress the cc_value as this
+			 * if condition expects that cc_value will change after each call to
+			 * set_timeout function.
+			 */
+			cyc = 1;
+		}
 
-	uint64_t delta_time = ticks * CYC_PER_TICK;
+		/* If it's the first timeout setting after previous expiration and timeout
+		 * is short so fast method can be used which utilizes relative CC configuration.
+		 */
+		cc_value += cyc;
+		nrfx_grtc_syscounter_cc_rel_set(ch, cyc, NRFX_GRTC_CC_RELATIVE_COMPARE);
+		return;
+	}
 
-	uint64_t target_time = counter() + delta_time;
+	uint64_t cyc = (uint64_t)ticks * CYC_PER_TICK;
+	bool safe_setting = false;
+	int64_t prev_cc_val = cc_value;
 
-	/* Rounded down target_time to the tick boundary
-	 * (but not less than one tick after the last)
+	cc_value = last_count + last_elapsed + cyc;
```

Comment:
I think the following will be clearer, and `now` is anyway needed a few lines below:
```suggestion
	int64_t now = last_count + last_elapsed;

	cc_value = now + cyc;
```